flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java

```diff
@@ -66,16 +71,36 @@ public ChangelogStateBackendHandleImpl(
 
         @Override
         public void registerSharedStates(SharedStateRegistry stateRegistry, long checkpointID) {
+            for (KeyedStateHandle keyedStateHandle : materialized) {
+                registerState(stateRegistry, checkpointID, MATERIALIZED_FLAG, keyedStateHandle);
+            }
+            for (ChangelogStateHandle stateHandle : nonMaterialized) {
+                registerState(stateRegistry, checkpointID, NON_MATERIALIZED_FLAG, stateHandle);
+            }
             stateRegistry.registerAll(materialized, checkpointID);
             stateRegistry.registerAll(nonMaterialized, checkpointID);
         }
 
+        private void registerState(
+                SharedStateRegistry stateRegistry,
+                long checkpointID,
+                String prefix,
+                KeyedStateHandle keyedStateHandle) {
+            stateRegistry.registerReference(
+                    new SharedStateRegistryKey(
+                            prefix,
+                            // here use hash code as registry key identifier.
+                            new StateHandleID(String.valueOf(keyedStateHandle.hashCode()))),
+                    new StreamStateHandleWrapper(keyedStateHandle),
+                    checkpointID);
```

Comment:
I'm concerned about the use of `hashCode` as a key in `SharedStateRegistry`:
1. According to java contract, it can be the same even for different state handles
2. State Backends are not required to implement this in the retuned handles
3. RocksDB does, but explicitly discourages from its use in production code:
`This method should only be called in tests! This should never serve as key in a hash map` (`IncrementalRemoteKeyedStateHandle` )

I see the following alternatives:
1. Materialization ID (I'm adding it in #18382 to `ChangelogStateBackendHandleImpl`)
2. SequenceNumber upTo/from for materialized/non-materialized parts
3. Random ID generated by backend and assigned to `ChangelogSnapshotState`.
(In 1 and 2, we'll need to add some backend specific part)

As long as the API is private, I'm not very concerned about the non-materialized part. Otherwise, SQN seems the best choice.

WDYT?

Reply:
Thanks for your advice. I think SequenceNumber is better for materialized/non-materialized part. However, these numbers do not persist and restart from `INITIAL_SQN` once restored. Do you think we could persist these numbers and reset the number from last restored?

Reply:
I think resetting Writer to a given SQN is possible with the current DFS changelog implementation, but this may be problematic in other implementations.
Using SQN also implies that the `changelogWriter.persist(sqn)` must return a new handle for each new SQN, even if nothting was changed (SQN can generally increment without state changes). Unless the returned handle doesn't contain any "private" state (which is currently the case), but then we don't need to worry about non-materialized state at all.

Reply:
Btw, won't it solve the problem if we register **all** files in `IncrementalRemoteKeyedStateHandle` with `SharedStateRegistry` (including metadata etc.)? This would be much simpler I think in all respects.
I guess previously there just was no need to register them (and they must be kept separate for restore).

Reply:
| won't it solve the problem if we register all files in IncrementalRemoteKeyedStateHandle with SharedStateRegistry (including metadata etc.)?

I think changelog keyed state backend should be designed to wrap any keyed state backend, we should consider the reference counting problem in changelog state backend instead of modify the state handle returned in the wrapped state backend. That's why I insist to change on changelog state-backend side.

Reply:
> I think resetting Writer to a given SQN is possible with the current DFS changelog implementation, but this may be problematic in other implementations.

If so, maybe leveraging the materialization id is better?

Reply:
Yes, I think materialization id is a way to go.

Reply:
Already refactored by leveraging materialization id to register.

Reply:
Thanks for updating the PR. Unfortunately, I have some further conerns regarding `SharedStateRegistryKey`.

The updated version uses `backendIdentifier + "-" + materializationID` as a key.

How would it handle the re-scaling case?
Per my understanding, `backendIdentifier` will be different than what was used on JM while reading theckpoints.
Additionally, on down-scaling, `materializationID` can change for some handles.

If that's a valid concern, we probably have to use optiton (3) random IDs per materalized handle, and store/pass them inside `ChangelogStateBackendHandleImpl ` maybe as a map.